[FIX] website: user navbar check for `main_object`

simahawk opened the pull request:
**Current behavior before PR:**

User navbar rendering broken if `main_object` is None.

**Desired behavior after PR is merged:**

Make it fault tolerant and check if that object is really there.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

simahawk commented:
@mart-e @JKE-be any feedback? :)

JKE-be commented:
Hi @simahawk 

In which case you don't have it? 

render call _prepare_qcontext that should always add it!
https://github.com/odoo/odoo/blob/b42b9c936b64dca72a41422e008537c34b691912/odoo/addons/base/ir/ir_ui_view.py#L1164

https://github.com/odoo/odoo/blob/47b376e66d2e9f303934a18800070bdb48d472d0/addons/website/models/ir_ui_view.py#L127

simahawk commented:
@JKE-be hi, there you go https://travis-ci.org/OCA/website-cms/jobs/336075786#L1137

Context: `main_object` is the name for the current record that you are editing and is generally used across all odoo's templates.

In that module that throws the error `main_object` is still the object that you can edit BUT it can be an object that you are creating (like a page, a partner, pick one model) and the controller will set `main_object = None`.

So, in such condition `main_object` is not there.

I could change that name but I'd like to stick to odoo's naming to make everybody happy (the publisher widget too).

The fix it's really trivial and secure, and is also a good practice AFAIK :)

JKE-be commented:
Not really, the bug comes from https://github.com/simahawk/website-cms/blob/9.0/cms_form/controllers/main.py#L109

Never in Odoo, you should have a main_object null.
you should have at least a browse record null IMHO if you don't have an id.

main_object = request.env[model] (instead of null)

simahawk commented:
> Never in Odoo, you should have a main_object null.

and why? just because of that template? Before that change it was never needed and it's the only place that requires it AFAIK.

I can live w/ the fix on my side but IMHO this is a wrong assumption.

simahawk commented:
Also, semantically an empty recordset and `None` are not the same thing.
Your check is not going to work fine anyway since `recset.fields_get()` is going to return `website_published` field and you gonna get the publisher even if you don't have an object. Which is wrong.

simahawk commented:
@JKE-be is that thumb-up a sign of rethinking? :smile:

JKE-be commented:
Sign that browse Record null was probably not the best solution :)
The current view if not main_object is the default behaviour.

But I continue to think that if I merge it, that will fix one of for your case, but you will have other bugs like meta SEO broken, ... Main Object should be "required" to render the view IMHO.

simahawk commented:
@JKE-be not at all. Here's where the main template use it:

https://github.com/odoo/odoo/blob/11.0/addons/website/views/website_templates.xml#L101

```xml
[...]
'data-main-object': repr(main_object) if editable or translatable else None,
[...]
```
which will give you 'None'.
Then it's retrieved by JS here 
https://github.com/odoo/odoo/blob/11.0/addons/website/static/src/js/menu/seo.js#L462
```javascript
    getMainObject: function () {
        var repr = $('html').data('main-object');
        var m = repr.match(/(.+)\((\d+),(.*)\)/);
        if (!m) {
            return null;
        } else {
            return {
                model: m[1],
                id: m[2]|0
            };
        }
    },
```
and is used safely, for instance:
```xml
    loadMetaData: function () {
        var obj = this.getMainObject();
        var def = $.Deferred();
        if (!obj) {
            // return $.Deferred().reject(new Error("No main_object was found."));
            def.resolve(null);
[...]
```
So, it's safe from what we can see.

Also, once again, you are assuming that displaying those actions/menu/whatever is correct even if you don't have anything to play with.

JKE-be commented:
> which will give you 'None'.

Not at all, that will return (ir_ui_view, <xml_id_of_view>) when you call render(xml_id_of_view, values) withtout force a main_object to None like in your code.

```python
            if 'main_object' not in qcontext:
                qcontext['main_object'] = self
```
where self is the current template.

But yes, in some case, we manage the case where 'get_main_object' is null, so we could eventually add your diff, but I continue to think that it is a bad idea toset a main_object to None ;)

simahawk commented:
> Not at all, that will return (ir_ui_view, <xml_id_of_view>) when you call render(xml_id_of_view, values) withtout force a main_object to None like in your code.
> 
>             if 'main_object' not in qcontext:
>                 qcontext['main_object'] = self

aaaaaand? :thinking: 

what I'm saying is that if you set `main_object = None` - and you do it consciously (hopefully :smile:) as a developer - you won't get any issue, because the std behavior won't break. 

> but I continue to think that it is a bad idea to set a main_object to None ;)

I already gave you a general use case where you don't have any object and the view fallback that you explained is fine.

Plus, talking about my specific case, it's useless - and dangerous from a certain POV - to show SEO things or the publisher if you don't have any record (create form) or if you are editing a record (edit form).
So, again, such use cases are perfectly safe and fine w/ `None`.

Finally, the change I proposed above won't break anything in your current code.

simahawk commented:
@JKE-be here's another good example:

1. I removed my fix
2. I've implemented the change you suggested (main_object defaults to empty recordset) [1]
3. I created a search form for events

![image](https://user-images.githubusercontent.com/347149/37713003-df4fdd58-2d15-11e8-859b-6358299e0d9f.png)

Getting that publisher widget here is totally wrong as there's nothing to publish/unpublish.

[1] this makes `website_event_track` happy since it has the same issue, as well as `website_sale`
/cc @jbeficent
